tools/generate_report.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3
+
+#############################################################################
+##
+##  This file is part of GAP, a system for computational discrete algebra.
+##
+##  Copyright of GAP belongs to its developers, whose names are too numerous
+##  to list here. Please refer to the COPYRIGHT file for details.
+##
+##  SPDX-License-Identifier: GPL-2.0-or-later
+##
+
+# Run this script on main branch with data and gh-pages worktree
+
+"""
+This script compares the current test-status.json with a previous version,
+generates a main report.md along with
+    a test-status-diff.json,
+    a html-redirect, and
+    a badge.
+"""
+
+from utils import warning, error, symlink, string_to_bool
+
+import sys
+import os
+import json
+
+################################################################################
+# Arguments and Paths
+num_args = len(sys.argv)
+
+if num_args < 1 or num_args > 4:
+    error('Unknown number of arguments')
+
+# relative paths to report directories from ROOT
+ROOT = 'data/reports'
+os.makedirs(ROOT, exist_ok = True)
+DIR_LAST_REPORT_REL = 'latest'
+OVERRIDE_LAST = True
+
+if num_args > 1: DIR_REPORT_REL = sys.argv[1]
+if num_args > 2: DIR_LAST_REPORT_REL = sys.argv[2]
+if num_args > 3: OVERRIDE_LAST = string_to_bool(sys.argv[3])
+
+DIR_REPORT = os.path.realpath(os.path.join(ROOT, DIR_REPORT_REL))
+DIR_LAST_REPORT_SYMBOLIC = os.path.join(ROOT, DIR_LAST_REPORT_REL)
+DIR_LAST_REPORT = os.path.realpath(DIR_LAST_REPORT_SYMBOLIC)
+
+REPORT_PATH = os.path.join(DIR_REPORT, 'test-status.json')
+LAST_REPORT_PATH = os.path.join(DIR_LAST_REPORT, 'test-status.json')
+
+if OVERRIDE_LAST:
+    DIR_BADGE = os.path.join('data/badges', DIR_LAST_REPORT_REL)
+    os.makedirs(DIR_BADGE, exist_ok = True)
+
+    DIR_REDIRECT = os.path.join('gh-pages', DIR_LAST_REPORT_REL)
+    os.makedirs(DIR_REDIRECT, exist_ok = True)
+
+################################################################################
+# Read current and previous test-status
+with open(REPORT_PATH, 'r') as f:
+    REPORT = json.load(f)
```

Comment:
I think we should add a function like this to `utils.py`, as we have similar code in soooo many places
```python
def read_json(path):
    with open(path, 'rb') as f:
        return json.load(f)
```
Note that I used `'rb`', to read as binary, with the hope that we can ignore encodings this way; but it may not work... Of course in the four places we call `json.load` right now, we do...
- `open(fname, 'r')`
- `open(fname, "rb")`
- `open(fname, "r", encoding="utf-8")`
- `open(fname, 'r', encoding='utf-8', errors='ignore')`

So actually unifying this may be difficult *sigh*. But at least within this file, we could do it...